### ISSUE TITLE
Parsing: Improving the handling of whitespace between block nodes

### ISSUE THREAD

**Issue by azaozz** [WordPress/gutenberg#34023]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Have you tried deactivating all plugins except Gutenberg?

- [X] I have tested with all plugins deactivated.

### Have you tried replicating the bug using a default theme e.g. Twenty Twenty?

- [X] I have tested with a default theme.

### Description

This is a follow-up to https://github.com/WordPress/gutenberg/pull/33136#issuecomment-874193764 (as the discussion is there but the PR was closed). 

Sorry but seems I'm missing something here :)

> This wouldn't be fixing a bug (empty nodes have been known and handled in Gutenberg since the beginning), it would be changing the actual specification of the block-serialisation format.

If I'm understanding the block-serialisation format correctly white space/line breaks between block boundaries are ignored in the block format, same as in HTML. If they are not, why are they being added to the saved content? Inserting line breaks between block boundaries on save and then creating fallback/freeform blocks from them would suggest incorrect use of the format, right? 

The expectation, imho, would be to match the editor output format with the input format, no discrepancies.

> Even though the resulting "empty" freeform nodes can seem awkward in these situations, they are correct.

Hmm, don't think that's correct. These "empty nodes"/line breaks were not added by the users. They were added programmatically on saving the content, and then treated differently on loading that content. That behaviour seems inconsistent. It becomes very apparent/buggy when the `core/freeform` block is unregistered and/or the `FreeformContentHandlerName` is undefined.

Going to reopen this for a "second opinion", also cc @dmsnell.

### Step-by-step reproduction instructions

N/A

### Expected Behavior

N/A

### Current Behavior

N/A

**Comment by dmsnell** [comment]
I don't have strong feelings on this. There seems to be space to make a practical concession for the cases where the editor intentionally adds newlines on its output. If I remember further back decisions were made to make the output HTML pretty (i.e. easier to look at, copy, and edit from the text view) which is the reason for the discrepancy.

Technically I probably agree with @mcsf that this is not an issue for the parsing layer since it's more of a subjective call what to do with it. Even if the editor were to stop adding those newlines it doesn't mean someone might not intentionally add them, so we still have to deal with the question.

One idea is that we could add a default filter that removes empty freeform blocks. If someone wants to preserve the specific empty blocks then they can remove the filter, but otherwise it's probably just the editor printing out noise. I doubt it would cause any noticeable performance impact, and if it were to do so, I'm certain we could adjust the filter to mitigate it.

**Comment by azaozz** [comment]
> Even if the editor were to stop adding those newlines it doesn't mean someone might not intentionally add them

Right. If somebody adds them, that would be "user intention" and blocks should be created. The problem is mostly that there's a discrepancy between saving and loading of the content. For example, saving a post with three paragraph blocks will produce markup with three blocks in it. However loading that post in the editor will produce five blocks, three paragraphs and two freeform. This is an inconsistency that should probably be fixed/corrected. 

Then, if the freeform block is not present, these turn into `undefined` blocks. If the FreeformContentHandler is changed, it results in even more bugs, autop is applied, existing blocks are merged, etc. (but these are different issues).

**Comment by dmsnell** [comment]
@azaozz I agree, and I feel like you restated my suggestion from a different perspective. My suggestion is that we fix this more or less where it's created: in the editor or through a WP filter, not by changing how we parse and interpret the document.

**Comment by mcsf** [comment]
> > Even though the resulting "empty" freeform nodes can seem awkward in these situations, they are correct.

> Hmm, don't think that's correct. These "empty nodes"/line breaks were not added by the users. They were added programmatically on saving the content, and then treated differently on loading that content.

What I was claiming to be correct was the parser's decision to create empty freeform nodes upon consuming whitespace between demarcated blocks.

That said, yes, it can be surprising that saving three content blocks results in an output that will be parsed as five block nodes. However, it should be clear that we are dealing with different objects from different layers: the three saved blocks are blocks, while the five parsed are block _nodes_.

If we can make a clumsy analogy with the DOM, blocks (as returned by `wp.blocks.parse`) are `Element` instances, and block nodes (as returned by `wp.blockSerializationDefaultParser.parse`) are `Node` instances. A `Node` is a broader and less featureful category which includes text nodes, even empty ones. `Element` builds on top of it — just like the block API provides full attribute sourcing, validation, etc. on top of raw block nodes.

Anyway, coming back to:

> However loading that post in the editor will produce five blocks, three paragraphs and two freeform. This is an inconsistency that should probably be fixed/corrected.
> 
> Then, if the freeform block is not present, these turn into undefined blocks. If the FreeformContentHandler is changed, it results in even more bugs, autop is applied, existing blocks are merged, etc. (but these are different issues).

Addressing this at the saving level, as Dennis suggests, is not off the table — but before we spend time on this I'd like to look at concrete issues arising from this handling of whitespace. Are there tracked issues, or can you provide specific steps?

**Comment by azaozz** [comment]
> I agree, and I feel like you restated my suggestion...

Yep, sorry, was just trying to wrap my head around it from all possible sides.

> it can be surprising that saving three content blocks results in an output that will be parsed as five block nodes. However, it should be clear that we are dealing with different objects from different layers: the three saved blocks are blocks, while the five parsed are block nodes.

Right, I understand, thanks for explaining. Thinking perhaps the only thing left here would be to add something like this explanation to the parser docs.

> Addressing this at the saving level, as Dennis suggests, is not off the table.

Yes but removing line breaks between blocks will reduce the readability of the content as saved in the DB. People still look directly there sometimes. Don't have strong feelings about it but maybe good to keep the pretty formatting if possible.

> Are there tracked issues

#33832 was caused indirectly by this. The problem (as far as I understand) was that when there is no default Freeform block, empty, non-registered blocks (`wp:null` and/or `wp:undefined`) are being created from the white space but not removed later on. 

This was fixed by making the `core/html` block the Freeform content handler in #33164, however that results in several other bugs like removing existing HTML blocks when saving, applying autop on loading, etc. (I'll open issues for these too, was just following the "chain" to make sure low level "freeform" handling works as expected).

Leaving this open just for the parser docs change.

**Comment by dmsnell** [comment]
It makes sense to me to filter this _somewhere_. The PHP parser is already filterable:

https://github.com/WordPress/WordPress/blob/2f6eb9d25e4d63ab3c9c6dc58f85c113e9bf0442/wp-includes/blocks.php#L883

I feel like we could add a new JS filter here…

https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/src/api/parser/index.js#L153

…or at the end of or around that function so that we could reject or modify blocks. `parse()` already checks if the returned block is falsey and drops it if it is, so it wouldn't be a stretch to modify the result of `parseRawBlock()` (or use a pre-filter so we don't do all the unregistered block stuff in there, causing these headaches)

**Comment by mcsf** [comment]
> Thinking perhaps the only thing left here would be to add something like this explanation to the parser docs.

Where in the docs would be best? There's a few parsing-related ones: overview documents like `data-flow.md`, and specific documentation for different parsers:

* block-serialization-default-parser (html to raw block nodes)
* block-serialization-spec-parser (idem)
* block-api (raw block nodes to proper blocks)

**Comment by mcsf** [comment]
> Thinking perhaps the only thing left here would be to add something like this explanation to the parser docs.

Where in the docs would be best? There's a few parsing-related ones: overview documents like `data-flow.md`, and specific documentation for different parsers:

* block-serialization-default-parser (html to raw block nodes)
* block-serialization-spec-parser (idem)
* block-api (raw block nodes to proper blocks)

**Comment by shixianli158** [comment]
Where in the docs would be best? There's a few parsing-related ones: overview documents like data-flow.md, and specific documentation for different parsers:

block-serialization-default-parser (html to raw block nodes)
block-serialization-spec-parser (idem)
block-api (raw block nodes to proper blocks)

**Comment by azaozz** [comment]
> Where in the docs would be best?

Perhaps something short like:

`Note that empty space between block delimiters is treated as non-block HTML content and will produce "freeform" blocks.`

to /packages/block-serialization-default-parser/README.md.

Then a bit more details, perhaps something along these lines:

`Inserting line breaks between block delimiters when saving (to prettify the output HTML) will produce empty block nodes when the content is parsed. For example: using the default configuration and saving a post containing three Paragraph blocks will produce five block nodes when parsed - three Paragraph block nodes and two empty Freeform block nodes. The Freeform block nodes are removed later. The type of the empty block nodes depends on the FreeformContentHandler and UnregisteredTypeHandler settings and whether these blocks are registered.`

Just not sure where that should go. Seems the logical place would be /packages/blocks/README.md but it doesn't mention anything about blocks, block nodes, or parsing. The general /packages/editor/README.md talks about parsing and block nodes, perhaps it will be better there?
